api/api.py

```diff
@@ -7,210 +7,266 @@
-        webapp2.Route(r'/add',              JobsHandler, handler_method='add', methods=['POST']),
-        webapp2.Route(r'/<:[^/]+>',         JobHandler,  name='job'),
-        webapp2.Route(r'/<:[^/]+>/config.json', JobHandler,  name='job', handler_method='get_config'),
-        webapp2.Route(r'/<:[^/]+>/retry',   JobHandler,  name='job', handler_method='retry', methods=['POST']),
-    ]),
-    webapp2.Route(r'/api/gears',             GearsHandler),
-    webapp2_extras.routes.PathPrefixRoute(r'/api/gears', [
-        webapp2.Route(r'/<:[^/]+>',            GearHandler),
-        webapp2.Route(r'/<:[^/]+>/invocation', GearHandler, handler_method='get_invocation'),
-        webapp2.Route(r'/<:[^/]+>/suggest/<:[^/]+>/<:[^/]+>', GearHandler, handler_method='suggest'),
+def route(path, target, h=None, m=None, name=None):
+
+    # https://webapp2.readthedocs.io/en/latest/api/webapp2.html#webapp2.Route
+    return webapp2.Route(
+        # re.compile(path)
+        path.format(**routing_regexes),
+        target,
+        handler_method=h,
+        methods=m,
+        name=name
+    )
+
+def prefix(path, routes):
+
+    # https://webapp2.readthedocs.io/en/latest/api/webapp2_extras/routes.html#webapp2_extras.routes.PathPrefixRoute
+    return webapp2_extras.routes.PathPrefixRoute(
+        path.format(**routing_regexes),
+        routes
+    )
+
+endpoints = [
+    route('/api',                  root.Root),
+    prefix('/api', [
+
+        # System configuration
+
+        route('/config',           Config,              m=['GET']),
+        route('/config.js',        Config,  h='get_js', m=['GET']),
+        route('/version',          Version,             m=['GET']),
+
+
+        # General-purpose upload & download
+
+        route('/download',                              Download, h='download',              m=['GET', 'POST']),
+        route('/upload/<strategy:label|uid|uid-match>', Upload,   h='upload',                m=['POST']),
+        route('/clean-packfiles',                       Upload,   h='clean_packfile_tokens', m=['POST']),
+        route('/engine',                                Upload,   h='engine',                m=['POST']),
+
+
+        # Top-level endpoints
+
+        route('/resolve',                           ResolveHandler, h='resolve', m=['POST']),
+        route('/schemas/<schema:{schema}>',         SchemaHandler,               m=['GET']),
+        route('/report/<report_type:site|project>', ReportHandler,               m=['GET']),
+
+
+        # Search
+
+        route('/search',                     SearchHandler, h='advanced_search', m=['POST']),
+        route('/search/files',               SearchHandler, h='get_datatree',    m=['GET']),
+        route('/search/<cont_name:{cname}>', SearchHandler,                      m=['GET']),
+
+
+        # Users
+
+        route( '/users',                   UserHandler, 'get_all', m=['GET']),
```

Comment:
Just missing the `h` here, otherwise looks good! Because python it will work without it, though 😉.